### ISSUE TITLE
Do not update when data is the same

### ISSUE THREAD

**Issue by danziamo** [kolodny/immutability-helper#51]
It would be good if update function could return previous data if changed variable is the same

i mean 
`
let a = {b: {c: {d : 5}}};`
`let z = update(a, {b: {c: {d : {$set: 5}}});
`
here 'z' changes even though value is the same.
Right now i have a performance issue with react application, cause after update i receive new object most of the times with the same value as previous.

I believe Implementing this feature would speed up a react application

**Comment by kolodny** [comment]
Hi danziamo. This currently works out of the box unless I'm misunderstanding what you're referring to: 

https://runkit.com/embed/yofw29ccbyeq

```js
var update = require("immutability-helper")

let a = {b: {c: {d : 5}}};
let z = update(a, {b: {c: {d : {$set: 5}}}});
console.log(a === z) // true
```

**Comment by danziamo** [comment]
Hello kolodny, 

My bad
```
var update = require("immutability-helper")

let a = {b: {c: {d : [4, 5]}}};
let z = update(a, {b: {c: {d : {$set: [4, 5]}}}});
console.log(a === z) // false
```
I have a function that generates array almost every time in my application and i usually update an object with this generated array.
I understand that it will be a new object every time, even though it has the same values inside,
Are there any options not to update main object by making deepEqual of current value and new value?

Thank you

**Comment by drom** [comment]
@danziamo @kolodny I guess the right solution would be:

```js
var update = require("immutability-helper")

let a = {b: {c: {d : [4, 5]}}};
let z = update(a, {b: {c: {d: {'0': {$set: 4}, '1': {$set: 5}}}}});
console.log(a === z) // true
```

**Comment by kolodny** [comment]
Hmm, I'm playing with the idea of doing a shallow equal when checking for a change in the value. There's some perfomance hit but I'd imagine setting the value to an object is an edge case

b1b8a69

Thoughts?

**Comment by drom** [comment]
@kolodny We have cases when DeepEqual check is highly desirable. Could we implement another command `$setDeep` or something. That would perform DeepEquality check always?

**Comment by danziamo** [comment]
@kolodny 

I have a case like this, and i think shallow equals would not work, but deepEquals will.

```javascript
var a = {b: {c: {d: [{e: 4}, {f: 5}, {g: 6}]}}};
var z = {b: {c: {d: {$set: [{e: 4}, {f: 5}, {g: 6}]}}}};
console.log (a === z); //false
```

**Comment by kolodny** [comment]
Hmm, I don't want to default to a deepEqual. What about exposing an escape hatch for this? See 02243b7 for details.

**Comment by drom** [comment]
@kolodny I love escape hatches! :1st_place_medal: especially the one from LOST

**Comment by danziamo** [comment]
@kolodny Yes, very good idea. Thank you

**Comment by kolodny** [comment]
Published to `2.3.0`

**Comment by danziamo** [comment]
Thank you
